Fix CSP nonce validation and violation reporting for external scripts

dyegoaurelio opened the pull request:
This PR fixes two related issues with Content Security Policy (CSP) nonce validation for external scripts:

1. Missing nonce validation for external scripts with malformed attributes
2. Incorrect violation event reporting for blocked external resources


This makes servo closer to passing the `nonce-enforce-blocked` wpt test.

The remaining failures are blocked by required changes in the html parser.

1. Svg script support (https://github.com/servo/html5ever/issues/118)
```html
<svg xmlns="http://www.w3.org/2000/svg">
<script attribute attribute nonce="abc">
    t.unreached_func("Duplicate attribute in SVG, no execution.")();
</script>
</svg>
```

2. Duplicate attrs check
the html parser needs to provide this flag, as mentioned on the original commit message (https://github.com/servo/servo/commit/4821bc0ab01e1ed0bb27e86c2df545019bd3856a)

```html
<script attribute attribute nonce="abc">
    t.unreached_func("Duplicate attribute, no execution.")();
</script>
<script attribute attribute=<style nonce="abc">
    t.unreached_func("2# Duplicate attribute, no execution.")();
</script>

[...]

<script src="../support/nonce-should-be-blocked.js?5" attribute attribute nonce="abc"></script>
```

I've also created a PR to implement the duplicate attrs flag on html5ever https://github.com/servo/html5ever/pull/695

Testing: doesn't fixes the aforementioned wpt test yet.
Fixes: part of #36437

dyegoaurelio commented:
marked as draft because I created a couple wpt failures https://github.com/dyegoaurelio/servo/actions/runs/19787946509

TimvdLippe commented on components/script/dom/security/csp.rs:
```diff
@@ -279,30 +279,47 @@ impl GlobalCspReporting for GlobalScope {
             };
+            // Determine source location based on violation type and whether position was provided
+            let (source_file, line_number, column_number) =
+                if source_position_was_provided || is_inline {
```
Shouldn't these be both true? E.g. the source position should have only been provided if it was inline. Therefore, I think we should assert that if the resource is URL, no source position is provided. Then only in the URL case we set the filename and for the other ones we use the source position.

TimvdLippe reviewed (changes requested):
Nice fixes! I do think we can improve the handling the source position since the is_inline boolean seems duplicate information to me.

dyegoaurelio commented on components/script/dom/security/csp.rs:
```diff
@@ -279,30 +279,45 @@ impl GlobalCspReporting for GlobalScope {
+                    source_position.column_number,
+                )
+            } else {
+                (self.get_url().to_string(), 0, 0)
```
Makes sense. Thanks for the review!

> It could be good to write a WPT test that only tests for one thing (e.g. the URL without line and column number) and then we defer passing the other tests until the HTML parser is updated.

I didn't realize I could just add a new wpt test. I thought they were strictly pulled from upstream. 
I agree that it's much better this way. I'll add a new test


> If I understand you correctly, in the following > cases we should set the filename, but not the line numbers:
>
>    Link elements
>    Image elements

TBH, I'm not sure about the correct behavior here.

All the changes I made on this regard were because of test assertions:

1. I've removed the line number from the script sources because https://github.com/servo/servo/blob/c4db781e1784a2d28cd9e7e2c03482ee7f2b3aae/tests/wpt/tests/content-security-policy/script-src/nonce-enforce-blocked.html#L29-L37 expects line number not to be truthy in order for it to check the `blockedURI`

2. after your feedback I reworked it to not allow url resources (https://github.com/servo/servo/pull/40956/commits/eb1d86ac85abe24c9e2775a8624d11ec87298f59). But then It broke https://github.com/servo/servo/blob/c4db781e1784a2d28cd9e7e2c03482ee7f2b3aae/tests/wpt/tests/content-security-policy/reporting-api/reporting-api-sends-reports-on-violation.https.sub.html#L26-L41 Because this one expects _both_ to be defined. That's why I reverted the assertion and added the line numbers back for the img reports.

Also, I just noticed that the img test is checking the event `body.blockedURL` while  the script one is checking `e.blockedURI`

**Maybe** they require slightly different handling. I'm not sure.

I'll try to read the spec more thoughtfully later to get a better understanding of the correct way to build these objects.

TimvdLippe replied:
Thanks for the investigation! Note that for the second test, only Chrome passes it: https://wpt.fyi/results/content-security-policy/reporting-api/reporting-api-sends-reports-on-violation.https.sub.html?label=experimental&label=master&aligned

If we can't figure out a reasonable behaviour that passes all tests, we can also file an issue with the spec. For example, I am confused why that line expectation is 53, yet that's the line that has the script tag. Safari doesn't report a line number at all. Shouldn't it be line 49?

dyegoaurelio committed:
csp: Apply nonce nonceability check for external script fetches

External `<script>` elements with malformed attributes (e.g.,
`<script src="evil.js" <script nonce="abc">`) bypass the "is element
nonceable" check. Step 24 of "prepare the script element" was reading
`[[CryptographicNonce]]` directly via `nonce_value()`, skipping the
malformed-attribute check.

Add a nonceability guard at step 24: if the element has a nonce content
attribute but is not nonceable (malformed attributes per CSP §3.5),
strip the nonce. Elements without a nonce content attribute (e.g.
JS-created via `.nonce = "abc"`) use the internal slot directly, since
the nonceable check only applies to parser-created elements.

Spec: https://www.w3.org/TR/CSP/#is-element-nonceable

Signed-off-by: Dyego Aurélio <dyegoaurelio@gmail.com>

dyegoaurelio committed:
csp: Fix source position in violation reports for external resource fetches

Fix compute_scripted_caller_source_position to return column 0 (not 1)
when no script is on the stack, by matching on Ok/Err from
describe_scripted_caller instead of using unwrap_or_default which
unconditionally applied the +1 offset.

Remove explicit source_position from script, link, and preload fetch
response CSP handlers. Image element keeps its source_position to
match chrome behavior asserted by report-to-directive-allowed-in-meta.https.sub.html.

Also fix the Convert<CSPViolationReportBody> impl to conditionalize
sourceFile/lineNumber/columnNumber on source_file being non-empty,
matching the existing report-uri serialization.

Spec: https://www.w3.org/TR/CSP/#create-violation-for-global

Signed-off-by: Dyego Aurélio <dyegoaurelio@gmail.com>

dyegoaurelio committed:
wpt: Add servo-specific nonce enforcement test and update expectations

Add a custom servo WPT test that verifies nonce enforcement for both
inline and external scripts with malformed attributes. The test covers
the subset of nonce-enforce-blocked.html cases that Servo can currently
enforce without html5ever changes (duplicate-attribute detection, SVG
script support).

Remove expected failure for trusted-types/navigate-to-javascript-url-001
which now passes due to the column number fix in
compute_scripted_caller_source_position.

Signed-off-by: Dyego Aurélio <dyegoaurelio@gmail.com>

TimvdLippe reviewed (approved):
Nice work, thanks for cleaning this up!
